File: cpp/src/distinct_count_accuracy_profile.cpp
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

#include <iostream>
#include <algorithm>
#include <cmath>
#include <chrono>
#include <ctime>

#include "distinct_count_accuracy_profile.hpp"

namespace datasketches {

accuracy_stats::accuracy_stats(size_t k, size_t true_value):
true_value(true_value),
sum_est(0),
sum_rel_err(0),
sum_sq_rel_err(0),
count(0),
rel_err_distribution(k)
{}

void accuracy_stats::update(double estimate) {
  sum_est += estimate;
  const double relative_error = estimate / true_value - 1.0;
  sum_rel_err += relative_error;
  sum_sq_rel_err += relative_error * relative_error;
  rel_err_distribution.update(relative_error);
  count++;
}

size_t accuracy_stats::get_true_value() const {
  return true_value;
}

double accuracy_stats::get_mean_est() const {
  return sum_est / count;
}

double accuracy_stats::get_mean_rel_err() const {
  return sum_rel_err / count;
}

double accuracy_stats::get_rms_rel_err() const {
  return sqrt(sum_sq_rel_err / count);
}

size_t accuracy_stats::get_count() const {
  return count;
}

std::vector<double> accuracy_stats::get_quantiles(
    const double* fractions, size_t size) const {
  return rel_err_distribution.get_quantiles(fractions, size);
}

/**
 * Manages multiple trials for measuring accuracy.
 *
 * <p>An accuracy trial is run along the distinct count axis (X-axis) first. A single trial
 * consists of a single sketch being updated with the max distinct values, stopping at the
 * configured X-axis points along the way where the accuracy is recorded
 * into the accuracy_stats array. Each instance of accuracy_stats retains the distribution of
 * the relative error and some other measurements for all the trials at that X-axis point.
 *
 * <p>Because accuracy trials take a long time, this profile will output intermediate
 * accuracy results starting after min_trials and then again at trial intervals
 * determined by tppo until max_trials. This allows to stop the testing at
 * any intermediate trials point if sufficient accuracy is achieved.
 */
void distinct_count_accuracy_profile::run() {
  const size_t lg_min_trials = 4;
  const size_t lg_max_trials = 16;
  const size_t trials_ppo = 4;
  const bool print_intermediate = true; // print intermediate data

  const size_t minT = 1 << lg_min_trials;
  const size_t max_trials = 1 << lg_max_trials;

  const size_t lg_min_counts = 0;
  const size_t lg_max_counts = 32;
  const size_t counts_ppo = 16;

  const size_t quantiles_k = 10000;

  const size_t num_points = count_points(lg_min_counts, lg_max_counts, counts_ppo);
  size_t p = 1 << lg_min_counts;
  for (size_t i = 0; i < num_points; i++) {
    stats.push_back(accuracy_stats(quantiles_k, p));
    p = pwr_2_law_next(counts_ppo, p);
  }

  key = 0;

  const auto start_time = std::chrono::system_clock::now();

  // this will generate a table of data up to each intermediate number of trials
  size_t last_trials = 0;
  while (last_trials < max_trials) {
    const size_t next_trials = (last_trials == 0) ? minT : pwr_2_law_next(trials_ppo, last_trials);
    const int delta = next_trials - last_trials;
    for (int i = 0; i < delta; i++) {
      run_trial();
    }
    last_trials = next_trials;

    if (print_intermediate or next_trials == max_trials) {
      print_stats();
    }

    std::cout << "Cum Trials             : " << last_trials << std::endl;
    std::cout << "Cum Updates            : " << key << std::endl;
    const auto current_time = std::chrono::system_clock::now();
    const std::chrono::milliseconds cum_time_ms =
        std::chrono::duration_cast<std::chrono::milliseconds>(current_time - start_time);
    std::cout << "Cum Time, ms           : " << cum_time_ms.count() << std::endl;
    const double time_per_trial_ms = (cum_time_ms.count()) / last_trials;
    std::cout << "Avg Time Per Trial, ms : " << time_per_trial_ms << std::endl;

    const auto current_time_t = std::chrono::system_clock::to_time_t(current_time);
    std::cout << "Current time           : " << std::ctime(&current_time_t);

    const auto time_to_complete_ms = std::chrono::duration_cast<std::chrono::milliseconds>(
        cum_time_ms / last_trials * (max_trials - last_trials));
    const auto est_completion_time = std::chrono::system_clock::to_time_t(current_time + time_to_complete_ms);
    std::cout << "Est Time of Completion : " << std::ctime(&est_completion_time);

    std::cout << std::endl;
  }
}

void distinct_count_accuracy_profile::print_stats() const {
  for (const auto& stat: stats) {
    std::cout << stat.get_true_value() << "\t";
    std::cout << stat.get_count() << "\t";
    std::cout << stat.get_mean_est() << "\t";
    std::cout << stat.get_mean_rel_err() << "\t";
    std::cout << stat.get_rms_rel_err() << "\t";
    // quantiles
    const auto quants = stat.get_quantiles(FRACTIONS, FRACT_LEN);
    for (size_t i = 0; i < FRACT_LEN; i++) {
      const double quantile = quants[i];
      std::cout << quantile;
      if (i != FRACT_LEN - 1) std::cout << "\t";
    }
    std::cout << std::endl;
  }
}

}

File: go/distinct_count_accuracy_profile.go
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package main

import (
	"fmt"
	"math"
	"strings"
	"time"

	"github.com/apache/datasketches-go/common"
	"github.com/apache/datasketches-go/hll"
	"github.com/apache/datasketches-go/kll"
)

type DistinctCountAccuracyProfile struct {
	config    distinctCountJobConfigType
	sketch    hll.HllSketch
	stats     []baseAccuracyStats
	startTime int64
}

type accuracyStats struct {
	qsk         *kll.ItemsSketch[float64]
	sumLB3      float64
	sumLB2      float64
	sumLB1      float64
	sumUB3      float64
	sumUB2      float64
	sumUB1      float64
	sumEst      float64
	sumRelErr   float64
	sumSqRelErr float64
	trueValue   uint64
}

func NewDistinctCountAccuracyProfile(config distinctCountJobConfigType, tgtType hll.TgtHllType) *DistinctCountAccuracyProfile {
	sketch, _ := hll.NewHllSketch(config.lgK, tgtType)
	return &DistinctCountAccuracyProfile{
		config:    config,
		sketch:    sketch,
		stats:     buildLog2AccuracyStatsArray(config.lgMinU, config.lgMaxU, config.uppo, config.lgQK),
		startTime: time.Now().UnixMilli(),
	}
}

func newAccuracyStats(k int, trueValue uint64) *accuracyStats {
	qsk, _ := kll.NewKllItemsSketch[float64](uint16(k), 8, common.ItemSketchDoubleComparator(false), common.ItemSketchDoubleSerDe{})
	return &accuracyStats{
		qsk:       qsk,
		trueValue: trueValue,
	}
}

func (d *DistinctCountAccuracyProfile) run() {
	minT := 1 << d.config.lgMinT
	maxT := 1 << d.config.lgMaxT
	maxU := 1 << d.config.lgMaxU

	vIn := uint64(0)

	// This will generate a table of data for each intermediate Trials point
	lastTpt := 0
	for lastTpt < maxT {
		nextT := lastTpt
		if lastTpt == 0 {
			nextT = minT
		} else {
			nextT = int(pwr2SeriesNext(d.config.tppo, uint64(lastTpt)))
		}
		delta := nextT - lastTpt
		for i := 0; i < delta; i++ {
			vIn = d.runTrial(vIn)
		}
		lastTpt = nextT
		sb := &strings.Builder{}
		if nextT < maxT {
			if d.config.interData {
				sb.Reset()
				d.setHeader(sb)
				d.process(lastTpt, sb)
				fmt.Println(sb.String())
			}
		} else {
			sb.Reset()
			d.setHeader(sb)
			d.process(lastTpt, sb)
			fmt.Println(sb.String())
		}

		fmt.Printf("Config:             : %+v\n", d.config)
		fmt.Printf("Cum Trials          : %d\n", lastTpt)
		fmt.Printf("Cum Updates         : %d\n", vIn)
		currentTime_mS := time.Now().UnixMilli()
		cumTime_mS := currentTime_mS - d.startTime
		fmt.Printf("Cum Time            : %s\n", time.Duration(cumTime_mS*1000*1000))
		timePerTrial_mS := float64(cumTime_mS) / float64(lastTpt)
		avgUpdateTime_ns := timePerTrial_mS * 1e6 / float64(maxU)
		fmt.Printf("Time Per Trial, mSec: %f\n", timePerTrial_mS)
		fmt.Printf("Avg Update Time, nSec: %f\n", avgUpdateTime_ns)
		fmt.Printf("Date Time           : %s\n", time.Now().Format(time.RFC3339))
		timeToComplete_mS := int64(timePerTrial_mS * float64(maxT-lastTpt))
		fmt.Printf("Est Time to Complete: %s\n", time.Duration(timeToComplete_mS*1000*1000))
		fmt.Printf("Est Time at Completion: %s\n", time.Now().Add(time.Duration(timeToComplete_mS*1000*1000)).Format(time.RFC3339))
		fmt.Println("")
	}
}

func (d *DistinctCountAccuracyProfile) process(cumTrials int, sb *strings.Builder) {
	points := len(d.stats)
	for pt := 0; pt < points; pt++ {
		q := d.stats[pt].(*accuracyStats)

		trueUniques := q.trueValue
		meanEst := q.sumEst / float64(cumTrials)
		meanRelErr := q.sumRelErr / float64(cumTrials)
		meanSqErr := q.sumSqRelErr / float64(cumTrials)
		normMeanSqErr := meanSqErr / (float64(trueUniques) * float64(trueUniques))
		rmsRelErr := math.Sqrt(normMeanSqErr)

		relLb3 := q.sumLB3/float64(cumTrials)/float64(trueUniques) - 1.0
		relLb2 := q.sumLB2/float64(cumTrials)/float64(trueUniques) - 1.0
		relLb1 := q.sumLB1/float64(cumTrials)/float64(trueUniques) - 1.0

		relUb1 := q.sumUB1/float64(cumTrials)/float64(trueUniques) - 1.0
		relUb2 := q.sumUB2/float64(cumTrials)/float64(trueUniques) - 1.0
		relUb3 := q.sumUB3/float64(cumTrials)/float64(trueUniques) - 1.0

		sb.WriteString(fmt.Sprintf("%d", trueUniques))
		sb.WriteString("\t")

		sb.WriteString(fmt.Sprintf("%e", meanEst))
		sb.WriteString("\t")

		sb.WriteString(fmt.Sprintf("%e", meanRelErr))
		sb.WriteString("\t")

		sb.WriteString(fmt.Sprintf("%e", rmsRelErr))
		sb.WriteString("\t")

		sb.WriteString(fmt.Sprintf("%d", cumTrials))
		sb.WriteString("\t")

		// Quantiles
		quants, _ := q.qsk.GetQuantiles(GAUSSIANS_3SD, true)
		for i := 0; i < len(quants); i++ {
			sb.WriteString(fmt.Sprintf("%e", quants[i]/float64(trueUniques)-1.0))
			sb.WriteString("\t")
		}

		// Bound averages
		sb.WriteString(fmt.Sprintf("%e", relLb3))
		sb.WriteString("\t")
		sb.WriteString(fmt.Sprintf("%e", relLb2))
		sb.WriteString("\t")
		sb.WriteString(fmt.Sprintf("%e", relLb1))
		sb.WriteString("\t")

		sb.WriteString(fmt.Sprintf("%e", relUb1))
		sb.WriteString("\t")
		sb.WriteString(fmt.Sprintf("%e", relUb2))
		sb.WriteString("\t")
		sb.WriteString(fmt.Sprintf("%e", relUb3))
		sb.WriteString("\n")

	}
}

func (d *DistinctCountAccuracyProfile) setHeader(sb *strings.Builder) string {
	sb.WriteString("TrueU")
	sb.WriteString("\t")
	sb.WriteString("MeanEst")
	sb.WriteString("\t")
	sb.WriteString("MeanRelErr")
	sb.WriteString("\t")
	sb.WriteString("RMS_RE")
	sb.WriteString("\t")
	sb.WriteString("Trials")
	sb.WriteString("\t")
	sb.WriteString("Min")
	sb.WriteString("\t")
	sb.WriteString("Q(.00135)")
	sb.WriteString("\t")
	sb.WriteString("Q(.02275)")
	sb.WriteString("\t")
	sb.WriteString("Q(.15866)")
	sb.WriteString("\t")
	sb.WriteString("Q(.5)")
	sb.WriteString("\t")
	sb.WriteString("Q(.84134)")
	sb.WriteString("\t")
	sb.WriteString("Q(.97725)")
	sb.WriteString("\t")
	sb.WriteString("Q(.99865)")
	sb.WriteString("\t")
	sb.WriteString("Max")
	sb.WriteString("\t")
	sb.WriteString("avgLB3")
	sb.WriteString("\t")
	sb.WriteString("avgLB2")
	sb.WriteString("\t")
	sb.WriteString("avgLB1")
	sb.WriteString("\t")
	sb.WriteString("avgUB1")
	sb.WriteString("\t")
	sb.WriteString("avgUB2")
	sb.WriteString("\t")
	sb.WriteString("avgUB3")
	sb.WriteString("\t")
	sb.WriteString("Max")
	return sb.String()
}

func (d *DistinctCountAccuracyProfile) runTrial(key uint64) uint64 {
	d.sketch.Reset()

	lastUniques := uint64(0)
	for _, ostat := range d.stats {
		stat := ostat.(*accuracyStats)
		delta := stat.trueValue - lastUniques
		for u := uint64(0); u < delta; u++ {
			key++
			d.sketch.UpdateUInt64(key)
		}
		lastUniques += delta
		est, _ := d.sketch.GetEstimate()
		lb3, _ := d.sketch.GetLowerBound(3)
		lb2, _ := d.sketch.GetLowerBound(2)
		lb1, _ := d.sketch.GetLowerBound(1)

		ub1, _ := d.sketch.GetUpperBound(1)
		ub2, _ := d.sketch.GetUpperBound(2)
		ub3, _ := d.sketch.GetUpperBound(3)

		stat.update(est, lb3, lb2, lb1, ub1, ub2, ub3)
	}

	return key
}

func (a *accuracyStats) update(
	est float64,
	lb3 float64,
	lb2 float64,
	lb1 float64,
	ub1 float64,
	ub2 float64,
	ub3 float64,
) {
	a.qsk.Update(est)
	a.sumLB3 += lb3
	a.sumLB2 += lb2
	a.sumLB1 += lb1
	a.sumUB1 += ub1
	a.sumUB2 += ub2
	a.sumUB3 += ub3
	a.sumEst += est
	a.sumRelErr += est/float64(a.trueValue) - 1.0
	erro := est - float64(a.trueValue)
	a.sumSqRelErr += erro * erro
}

func buildLog2AccuracyStatsArray(lgMin, lgMax, ppo, lgQK int) []baseAccuracyStats {
	qLen := countPoints(lgMin, lgMax, ppo)
	qArr := make([]baseAccuracyStats, qLen)
	p := uint64(1) << lgMin
	for i := 0; i < qLen; i++ {
		qArr[i] = newAccuracyStats(1<<lgQK, p)
		p = pwr2SeriesNext(ppo, p)
	}
	return qArr
}
